deepmd/dpmodel/descriptor/dpa1.py

```diff
@@ -0,0 +1,911 @@
+        input_r = dmatrix.reshape(-1, nnei, 4)[:, :, 1:4] / (
+            np.linalg.norm(
+                dmatrix.reshape(-1, nnei, 4)[:, :, 1:4], axis=-1, keepdims=True
+            )
+            + 1e-12
+        )
+        gg = self.dpa1_attention(
+            gg, nlist_mask, input_r=input_r, sw=sw
+        )  # shape is [nframes*nloc, self.neei, out_size]
+        # nfnl x ng x 4
+        gr = np.einsum("lni,lnj->lij", gg, rr)
+        gr /= self.nnei
+        gr1 = gr[:, : self.axis_neuron, :]
+        # nfnl x ng x ng1
+        grrg = np.einsum("lid,ljd->lij", gr, gr1)
+        # nf x nloc x (ng x ng1)
+        grrg = grrg.reshape(nf, nloc, ng * self.axis_neuron).astype(
+            GLOBAL_NP_FLOAT_PRECISION
+        )
+        # nf x nloc x (ng x ng1 + tebd_dim)
+        if self.concat_output_tebd:
+            grrg = np.concatenate([grrg, atype_embd.reshape(nf, nloc, -1)], axis=-1)
+        return grrg, gr[..., 1:], None, None, sw
+
+    def serialize(self) -> dict:
+        """Serialize the descriptor to dict."""
+        return {
+            "@class": "Descriptor",
+            "type": "dpa1",
+            "@version": 1,
+            "rcut": self.rcut,
+            "rcut_smth": self.rcut_smth,
+            "sel": self.sel,
+            "ntypes": self.ntypes,
+            "neuron": self.neuron,
+            "axis_neuron": self.axis_neuron,
+            "tebd_dim": self.tebd_dim,
+            "tebd_input_mode": self.tebd_input_mode,
+            "set_davg_zero": self.set_davg_zero,
+            "attn": self.attn,
+            "attn_layer": self.attn_layer,
+            "attn_dotr": self.attn_dotr,
+            "attn_mask": self.attn_mask,
+            "activation_function": self.activation_function,
+            "resnet_dt": self.resnet_dt,
+            "scaling_factor": self.scaling_factor,
+            "normalize": self.normalize,
+            "temperature": self.temperature,
+            "smooth_type_embedding": self.smooth,
+            "type_one_side": self.type_one_side,
+            "concat_output_tebd": self.concat_output_tebd,
+            # make deterministic
+            "precision": np.dtype(PRECISION_DICT[self.precision]).name,
+            "embeddings": self.embeddings.serialize(),
+            "attention_layers": self.dpa1_attention.serialize(),
+            "env_mat": self.env_mat.serialize(),
+            "type_embedding": self.type_embedding.serialize(),
+            "exclude_types": self.exclude_types,
+            "env_protection": self.env_protection,
+            "@variables": {
+                "davg": self.davg,
+                "dstd": self.dstd,
+            },
+            ## to be updated when the options are supported.
+            "trainable": True,
+            "spin": None,
+        }
+
+    @classmethod
+    def deserialize(cls, data: dict) -> "DescrptDPA1":
+        """Deserialize from dict."""
+        data = data.copy()
+        check_version_compatibility(data.pop("@version", 1), 1, 1)
```

Comment:
We don't need a default value, as it's a new class, and can assume that the version is always stored.